Add eg component to evaluate_shelter()

xoto10 opened the pull request:
Add an endgame component to the blockedstorm penalty so that the penalty applies more uniformly through the game.

STC 10+0.1 th 1 :
LLR: -2.95 (-2.94,2.94) [0.50,4.50]
Total: 94063 W: 21426 L: 21118 D: 51519
http://tests.stockfishchess.org/tests/view/5cd4605c0ebc5925cf04bf43

LTC 60+0.6 th 1 :
LLR: 2.95 (-2.94,2.94) [0.00,3.50]
Total: 188232 W: 32808 L: 32090 D: 123334
http://tests.stockfishchess.org/tests/view/5cd47d0a0ebc5925cf04c4fd

Refactored code with higher constant values gave a more convincing LTC result:
LLR: 2.96 (-2.94,2.94) [0.00,3.50]
Total: 30050 W: 5330 L: 5066 D: 19654
http://tests.stockfishchess.org/tests/view/5cd6a0000ebc5925cf050653

Further work; The values used have not been carefully tuned so no doubt they can be tuned more accurately. Also, perhaps other parts of the king safety calculation could benefit from using this.

Bench 3624368

xoto10 commented:
@protonspring and others - the code used here doesn't look particularly elegant, perhaps it can be made cleaner and at the same time, more efficient. Does anyone have any suggestions?

protonspring commented:
Why won't we just use a Score like everything else?

xoto10 commented:
It felt like I needed to manipulate the mg and eg values separately a lot,
so i thought that would lead to a lot of combining and splitting out of the
values, but I might be wrong. I coded it like this initially and stayed
with it.
I'll create a diff to return a score and see what that looks like ...

xoto10 commented:
I created a new branch [bstorm2](https://github.com/xoto10/stockfish-xoto10/compare/master...bstorm2) using a Score.

I couldn't get github to display the diff between the original pr code and the revised code like I can locally, so the diff is below. It's possible this is slightly more efficient although I don't think there's much in it, and perhaps it is a bit clearer than my original code. Should I run an stc test of one against the other somehow? Maybe a [-3,1] or [0,4] of the new against old to get an idea if there is any noticeable performance difference?

```
diff --git a/src/pawns.cpp b/src/pawns.cpp
index db89b77..7ff95d6 100644
--- a/src/pawns.cpp
+++ b/src/pawns.cpp
@@ -172,7 +172,7 @@ Entry* probe(const Position& pos) {
 /// penalty for a king, looking at the king file and the two closest files.
 
 template<Color Us>
-void Entry::evaluate_shelter(const Position& pos, Square ksq, Value& bonusMg, Value& bonusEg) {
+void Entry::evaluate_shelter(const Position& pos, Square ksq, Score& shelter) {
 
   constexpr Color     Them = (Us == WHITE ? BLACK : WHITE);
   constexpr Direction Down = (Us == WHITE ? SOUTH : NORTH);
@@ -183,6 +183,7 @@ void Entry::evaluate_shelter(const Position& pos, Square ksq, Value& bonusMg, Va
   Bitboard ourPawns = b & pos.pieces(Us);
   Bitboard theirPawns = b & pos.pieces(Them);
 
+  Value bonusMg = mg_value(shelter);
   Value safetyMg = (shift<Down>(theirPawns) & BlockSquares & ksq) ? Value(374) : Value(5);
   Value safetyEg = Value(0);
 
@@ -204,7 +205,7 @@ void Entry::evaluate_shelter(const Position& pos, Square ksq, Value& bonusMg, Va
   }
 
   if (safetyMg > bonusMg)
-      bonusMg = safetyMg, bonusEg = safetyEg;
+      shelter = make_score(safetyMg, safetyEg);
 }
 
 
@@ -227,17 +228,17 @@ Score Entry::do_king_safety(const Position& pos) {
   else while (pawns)
       minPawnDist = std::min(minPawnDist, distance(ksq, pop_lsb(&pawns)));
 
-  Value bonusMg = -VALUE_INFINITE, bonusEg = Value(0);
-  evaluate_shelter<Us>(pos, ksq, bonusMg, bonusEg);
+  Score shelter = make_score(-VALUE_INFINITE, Value(0));
+  evaluate_shelter<Us>(pos, ksq, shelter);
 
   // If we can castle use the bonus after the castling if it is bigger
   if (pos.can_castle(Us | KING_SIDE))
-      evaluate_shelter<Us>(pos, relative_square(Us, SQ_G1), bonusMg, bonusEg);
+      evaluate_shelter<Us>(pos, relative_square(Us, SQ_G1), shelter);
 
   if (pos.can_castle(Us | QUEEN_SIDE))
-      evaluate_shelter<Us>(pos, relative_square(Us, SQ_C1), bonusMg, bonusEg);
+      evaluate_shelter<Us>(pos, relative_square(Us, SQ_C1), shelter);
 
-  return make_score(bonusMg, -16 * minPawnDist + bonusEg);
+  return shelter - make_score(Value(0), 16 * minPawnDist);
 }
 
 // Explicit template instantiation
diff --git a/src/pawns.h b/src/pawns.h
index de8e6c3..771e9cd 100644
--- a/src/pawns.h
+++ b/src/pawns.h
@@ -50,7 +50,7 @@ struct Entry {
   Score do_king_safety(const Position& pos);
 
   template<Color Us>
-  void evaluate_shelter(const Position& pos, Square ksq, Value& bonusMg, Value& bonusEg);
+  void evaluate_shelter(const Position& pos, Square ksq, Score& shelter);
 
   Key key;
   Score scores[COLOR_NB];
```

protonspring commented:
I think this is a bit cleaner.

xoto10 commented:
Regarding the best values to use, an STC test [here](http://tests.stockfishchess.org/tests/view/5cd56d0e0ebc5925cf04dc30) gave nearly +1 Elo on top of this patch, suggesting 82,82 might be better values than 66,66. I leave it to maintainers to decide if we test those at ltc or we leave that for people to tune at a later date.

vondele commented:
@xoto10 what about testing the refactored patch with the new constants at LTC vs master. Will give a clean datapoint, now the PR is presenting at least 3 options.

xoto10 commented:
@vondele That seems like a good idea - started [here](http://tests.stockfishchess.org/tests/view/5cd6a0000ebc5925cf050653).

anshulongithub commented:
The new values passed at LTC with much fewer games than before

xoto10 commented:
As anshulongithub said, the new test was convincing, so I've updated to the refactored code, squashed and pushed to this PR, and updated the PR text with the details of the extra test.

mcostalba commented:
Merged in 5f4d44fda05333
